### ISSUE TITLE
Performance reduced after migrating from 4.5.0.FINAL to 5.3.1.RELEASE

### ISSUE THREAD

**Issue by j-razeed** [lettuce-io/lettuce-core#1345]
Application with spring-boot 1.5.9, spring-data-redis 1.8.9 and lettuce 4.5.0.FINAL. Redis-server version is 6.0.5 I did performance benchmark on my machine for records (40 KB) insertion and the throughput is 2857 records inserted per sec.

I upgraded the application to spring-boot 2.3.1, spring-data-redis 2.3.1 and lettuce 5.3.1.FINAL. Redis-server version is 6.0.5. I didn't change the code to create connection. Just upgraded the spring boot along with lettuce. Now when I do benchmark, the performance is reduced lot. i.e., 1000 records (same 40KB) inserted per second.

Do we need to any configuration change after migrating to latest version?

Refer the code to create connection factory (used in both version of my app) below :

```
@Bean
public LettuceConnectionFactory lettuceConnectionFactory()
{
    return new LettuceConnectionFactory(hostName, port);
}

@Bean
public RedisTemplate<String, Object> redisTemplate(LettuceConnectionFactory lettuceConnectionFactory) 
{
    RedisTemplate<String, Object> template = new RedisTemplate<>();
    template.setConnectionFactory(lettuceConnectionFactory);
    return template;
}
```

Below is the code used to benchmark the insertion

```
    long startTime = System.currentTimeMillis();
    Arrays.stream(persons).forEach(person -> {
        try {
            personRepository.save(person);
        } catch (Exception e)
        {
            log.error("Error is ", e);
        }
    });
    long timeElapsed = System.currentTimeMillis() - startTime;
    calculateThroughput(timeElapsed);
```
Where, persons is the array of Person objects. Throughput is calculated as below

```
 private void calculateThroughput(long timeElapsed)
    {
        long timeElapsedInSeconds = timeElapsed/1000;
        long recPerSec = noOfRecords/timeElapsedInSeconds;
        log.warn("It took {} milliseconds to persist {} number of records. i.e. {} records per second",
                timeElapsed, noOfRecords, recPerSec);
    }
```
Note that, I didn't change above logic after migrating to latest spring boot.

**Comment by mp911de** [comment]
This is a lot of Spring Data code that is Part of what you try to benchmark. Can you reproduce the change in performance behavior using plain Lettuce code?

If so, happy to discuss over an actual JMH benchmark.

**Comment by j-razeed** [comment]
Thanks for your clarification.

I did benchmarking with below plain lettuce code on versions 4.5.0.FINAL and 5.3.1.RELEASE.

RedisClient client = RedisClient.create("redis://localhost");

StatefulRedisConnection<String, String> connection = client.connect();

RedisCommands<String, String> commands = connection.sync();
commands.flushall();
long noOfRecords = 10000;
long startTime = System.currentTimeMillis();

LongStream.range(0L, noOfRecords)
		.forEach(value ->  commands.hset("key" + value, "id", String.valueOf(value)));

long timeElapsed = System.currentTimeMillis() - startTime;
calculateThroughput(timeElapsed, noOfRecords);

connection.close();

client.shutdown();

There is no change in performance i.e., got same result in both versions. So as you mentioned, no impact on plain lettuce.

We are planning to migrate our product to latest version of spring boot. So I did above benchmarking as a proof of concept. But the result is not happy. I am surprised to see more than 50% drop in throughput. Does latest version of spring-data-redis became heavy weight when compared to 1.8.9 version. Or Am I missing any configuration?
